crates/circuit/src/slice.rs

```diff
@@ -0,0 +1,375 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use thiserror::Error;
+
+use pyo3::exceptions::PyIndexError;
+use pyo3::prelude::*;
+use pyo3::types::PySlice;
+
+use self::sealed::{Descending, SequenceIndexIter};
+
+/// A Python-space indexer for the standard `PySequence` type; a single integer or a slice.
+///
+/// These come in as `isize`s from Python space, since Python typically allows negative indices.
+/// Use `with_len` to specialize the index to a valid Rust-space indexer into a collection of the
+/// given length.
+pub enum PySequenceIndex<'py> {
+    Int(isize),
+    Slice(Bound<'py, PySlice>),
+}
+
+impl<'py> FromPyObject<'py> for PySequenceIndex<'py> {
+    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
+        // `slice` can't be subclassed in Python, so it's safe (and faster) to check for it exactly.
+        // The `downcast_exact` check is just a pointer comparison, so while `slice` is the less
+        // common input, doing that first has little-to-no impact on the speed of the `isize` path,
+        // while the reverse makes `slice` inputs significantly slower.
+        if let Ok(slice) = ob.downcast_exact::<PySlice>() {
+            return Ok(Self::Slice(slice.clone()));
+        }
+        Ok(Self::Int(ob.extract()?))
+    }
+}
+
+impl<'py> PySequenceIndex<'py> {
+    /// Specialize this index to a collection of the given `len`, returning a Rust-native type.
+    pub fn with_len(&self, len: usize) -> Result<SequenceIndex, PySequenceIndexError> {
+        match self {
+            PySequenceIndex::Int(index) => {
+                let index = if *index >= 0 {
+                    let index = *index as usize;
+                    if index >= len {
```

Comment:
Could this also be written in a single line as follows? 🙂 
```suggestion
                    if *index as usize >= len {
```

Reply:
I don't remember exactly why I split it off like this, but if you want to change it, line 54 would also need changing to the same thing.  If you want to remake the suggestion with both I'll commit it.